### PR TITLE
Lexer skeleton

### PR DESCRIPTION
**Pull request by Nercury**
The idea here is to figure out how Lexer is going to look and work.

I placed all the modules into a single file, and when we finish the discussion I will move everything to appropriate files.

The `//` comments will be removed.

There are few decisions I made, quite possibly biased, so I will let you know my ideas:

First, I think it is possible to separate the Lexer from extensions completely, and initialize it with  collected list of options. The list is not big: just operators, and lexer does not even care if they are binary or unary.

Second, I want to keep the iterator with the idea of "lexing on demand". That is, lexing only happens when someone (parser) requests next token. However, this is internal detail, but good to keep in mind.

I want to try and keep only the slices to original string. What to do about new lines? Well, I suggest we think about that later :)

So, basically, this is a starting point. Won't merge until we are happy with result.

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -1,3 +1,162 @@
-#[test]
-fn it_works() {
+#[cfg(test)]
+mod test {
+    use tokens::{ Lexer, LexerOptions };
+
+    #[test]
+    fn lexer_usage() {
+        // build the lexer once for project environment with extensions.
+        let lexer = Lexer::new(LexerOptions::default(), vec![]);
```
Comment (colin-kiegel) [inline src/lib.rs]: I agree to with this reduced signature for the Lexer constructor. :-)

File: src/lib.rs [hunk 2]
```diff
@@ -1,3 +1,162 @@
...
+
+        // use many times.
+        {
+            let source: String = "{{ var }}".into();
+            for token in lexer.tokens(&source) {
```
Comment (colin-kiegel) [inline src/lib.rs]: I see the motivation behind the iterator pattern - however it has one drawback: It looses valuable information about 'current state'. The lexer definitely has different states. However you turn it, in a for-loop there will always be one entry-point `Iterator<>::next()` - each call to next() would need to lookup the last state and conditionally jump to the corresponding code (either vtable call or some match-pattern).

My current design does not use the iterator pattern for this reason. I don't need to carry a state at runtime, because it is all encoded statically. Each jump is without dynamic indirection or state-lookup-and-branching. And I might happen to generate more than one token between jumps. For an iterator design we would have to make frequent stops and dynamically store our state instead of static jumps.

This is a trade-off:
- fast static jumps means high troughput in terms of "lexed templates per second" for multiple templates - but you have to wait until a template is parsed as a whole
- iteration means you can build a lexer-parser-pipeline which can better parallelize lexing and parsing for a single template - but will use more cpu-cycles per template.

What will be important in the end? Ideally we will cache compiled templates and only regenerate if something changed (different template, different extensions, etc.).

Therefore I think it is not super important - but my tendency would be "static jumps - no iteration".
- increased throughput
- decreased latency per template regeneration (you know what I mean)

PS: If we are going to build a low-latency-pipeline this would still be possible without iterators. I read some interesting article about the "disruptor pattern" a while ago. So if we decide against iterator, we could still do something cool like that. :-)

File: src/lib.rs [hunk 3]
```diff
@@ -1,3 +1,162 @@
...
+                println!("{:?}", token);
+            }
+        }
+    }
+}
+
+// I don't know where to put this, keeping it in root for now.
+#[derive(Debug, Default, Clone)]
+pub struct Position {
```
Comment (colin-kiegel) [inline src/lib.rs]: In my design, the lexer has a helper object called "cursor" its supposed to keep track of the current position and do conversions like char-position to line:col. Maybe somewhere around there?

EDIT: Ok, I did use a struct like this in lexer::token::stream. (Previously stated that I didnt).

File: src/lib.rs [hunk 4]
```diff
@@ -1,3 +1,162 @@
...
+pub mod tokens {
+
+    // It is possible to delay all string manipulation for later,
+    // and we can simply store the slices into original full source
+    // string.
+    //
+    // We can keep doing that even for Node<'a>, provided the strings remain untouched.
+    // If something needs to be changed, we can create a special Node for that.
+    #[derive(Debug)]
+    pub enum TokenRef<'a> {
```
Comment (colin-kiegel) [inline src/lib.rs]: OK, cool :-)

File: src/lib.rs [hunk 5]
```diff
@@ -1,3 +1,162 @@
...
+pub struct Position {
+    pub line: usize,
+    pub column: usize,
+}
+
+// Named this "tokens", in plural, to mean a place where you should expect
+// to find your tokens.
+// Similar convention would work for "nodes", that's where AST lives, and possibly
+// "instructions", if we decide to go that route.
+pub mod tokens {
```
Comment (colin-kiegel) [inline src/lib.rs]: I am undecided whether I prefer tokens/nodes/templates or lexer/parser/runtime at top-level.

This largely depends on how it will feel like to use this library in the end. Maybe it's worth to take a look at other template engines and their public api? It's on my todo list, but I didn't have time yet.

Of course, we can start with one design and refactor later. At least for me, this will be some learning process.

File: src/lib.rs [hunk 6]
```diff
@@ -1,3 +1,162 @@
...
+        }
+    }
+
+    // This will be used when we need to carry token lifetime longer than original
+    // source string, for example, in error messages.
+    pub enum Token {
+        Text(String),
+    }
+
+    // Not pub, to make API more convenient.
```
Comment (colin-kiegel) [inline src/lib.rs]: good idea. I had almost everything public (not good). ^^

File: src/lib.rs [hunk 7]
```diff
@@ -1,3 +1,162 @@
...
+            pub fn new(options: Options, operators: Vec<&'static str>) -> Lexer {
+                // Here we will create patterns (I called them matchers), and
+                // store them in Lexer
+                Lexer
+            }
+
+            // twig-rust: https://github.com/colin-kiegel/twig-rust/blob/master/src/lexer/mod.rs#L64
+            // twig-rs: https://github.com/Nercury/twig-rs/blob/master/src/tokens/lexer/mod.rs#L40
+            //
+            // I think it is possible to avoid the Template::Raw in lexer API.
```
Comment (colin-kiegel) [inline src/lib.rs]: Yes, Template::Raw turned out to be quite close to useless. I was just giving everything names. On the other hand, it didn't hurt - so I kept it. :-)

File: src/lib.rs [hunk 8]
```diff
@@ -1,3 +1,162 @@
...
+            // We can probably deal with newlines in patterns?
+            // Also maybe we won't need to fix line endings, but right now we both do that.
```
Comment (colin-kiegel) [inline src/lib.rs]: agree - or we could make it an additional option.

File: src/lib.rs [hunk 9]
```diff
@@ -1,3 +1,162 @@
...
+            //
+            // twig-rs result was "Iter", twig-rust - "Job" :)
+            //
+            // I changed it to comcrete "Tokens" for now, which will implement Iterator.
+            // No Result. Let's avoid lexing until Parser requests first token.
```
Comment (colin-kiegel) [inline src/lib.rs]: Hmm ... let's return something "iterable".

long-term: disruptor pattern => low-latency + high-throughput
short-term: Vector => don't-talk-about-latency + high-throughput

I think this is a good compromise without the need to substantially refactor my lexer design. The disruptor is something similar to a ring-buffer.

What do you think?

**Comment by colin-kiegel** [comment]
I hope I didn't make too many comment. Awaiting your replies. ;-)

File: src/lib.rs (hunk 9 above)
Comment (colin-kiegel) [inline src/lib.rs]: PS: I continued thinking about this disruptor / concurrency pattern. If we go down this concurrency road, we should try to avoid to spawn and terminate threads too often. We might want to look into thread-pooling. There seems to be at least one rust-library for that: http://frewsxcv.github.io/rust-threadpool/threadpool/struct.ThreadPool.html

And there seems to be some implementations of the disruptor pattern for rust as well.

I might be able do some tests next weekend - possibly benchmarks, too.

File: src/lib.rs (hunk 2 above)
Comment (Nercury) [inline src/lib.rs]: OK, let's just lex everything on `tokens` call :)

Do we risk of blowing stack with your lexer? Or does Rust optimize tail calls in your case?

File: src/lib.rs (hunk 5 above)
Comment (Nercury) [inline src/lib.rs]: Assume we don't want to separate everything into separate `lexer`, `parser`, `node`, `token` (I tried, it is a bit annoying when things are all over the place).

So, a kind of "normalized" and "completely named" structure would have these names (which show the dependencies):

```
token
node
string_to_token_lexer
token_to_node_parser
node_to_whatever_compiler
```

So lexer is fine (can be either `lexer/token` or `token/node`).

But if then we create `lexer/token`, the parser would have to reach into the guts of lexer to get a token :)

Same kind of thing with compiler.

So, it would be nice to treat `lexer` and `parser` as kind of "detail" how those tokens and nodes are made :)

File: src/lib.rs (hunk 2 above)
Comment (colin-kiegel) [inline src/lib.rs]: Ok, cool. 

Good point about tail-calls. We will have to check whether tail-calls are optimized in this case by LLVM. Tail-calls seem not to be guaranteed natively by rust. It will depend on the good-will of LLVM and possibly compiler flags.
https://doc.rust-lang.org/complement-design-faq.html#no-guaranteed-tail-call-optimization
https://mail.mozilla.org/pipermail/rust-dev/2013-April/003557.html

I created an reminder #8 to investigate this.

File: src/lib.rs (hunk 5 above)
Comment (colin-kiegel) [inline src/lib.rs]: I think TwigPHP has chosen a very weird name "environment" ... what's that supposed to mean? Does it simulate global warming? No. ;-)

What TwigPHP calls "environment", I prefer to call compiler in a broad sense. With my Wikipedia-knowledge about compilers I would choose the naming like this: The compiler consists of a compiler-frontend (=lexer/parser, analyzing the source code - building an internal representation). What TwigPHP calls "compiler", I would suggest to call a compiler-backend (=optimizer/code generator) - again based on some Wikipedia-knowledge. ;-)

So, if I write compiler, I probably mean Twig-environment. What do you think about this renaming environment -> compiler? Does it make things clearer from a "consumers" point of view?

My motivation behind my crate-layout was this: The "consumer" of this library should be able to either
a) choose a ready-to-use compiler with default-configuration
b) or create a customized compiler by composition of its building blocks (custom lexer, custom parser, ..)

That's why I found it natural to make the "building blocks" of the compiler the "consumer-facing modules".

I think a consumer of this library, will probably not be very much interested about node representations etc., but more about how to configure and tweak the whole thing and its parts, injecting extensions here and there.

You are right, that the parser reaches into the guts of the lexer, in my current layout. I was considering to reverse this - it might make more sense if every part defines its input, something like "_you_ tell me what _you_ need _I_ tell you, what _I_ need"-principle instead of current "I tell you what you need and you tell me what I need"-principle (=lexer, parser, etc. talking with each other on their first date). But that would just revert the direction - then lexer would need to look into "details" of parser.

From the building-blocks perspective of a compiler - which I assume to be the perspective of a library consumer - parser / lexer / loader / runtime / ... are probably more important details than nodes / tokens / ... I am not 100% sure - just trying to take this perspective.

**Comment by Nercury** [comment]
Now I at looking at your `colin/pr_twig_skeleton` as an example, looks like we still using a bit different structure.

Notable differences:
I decided not to hide `lexer` inside `parser` for now (I know, changing my mind all the time :) ).
I am naming the container of `lexer`, `parser` the `api`. For now :/.
I would like `engine` with setup to be separate.
Any options the `lexer` needs should be passed by engine using `Options` object the lexer defines. Lexer will have no idea about `engine`.
I don't like macros in error handling :/. I will try to use them though.

**Comment by Nercury** [comment]
Right now I [implemented all lexer as iterator](https://github.com/rust-web/twig/commit/3d5cc2a5f4def1f070ce1d11ca0a2ee25b9014b6#diff-c568bdea77cb6e989977c775ab1efb89R33), but that does not prevent doing all lexing on first token request :).

Right now lexer is using `SyntaxError`/`SyntaxErrorCode` where `SyntaxErrorCode` contains just the message, and `SyntaxError` additionally has position in template. Note that here I have not used neither my `At` nor `ErrorCode` trait.

I am thinking about sharing `SyntaxError` between lexer and parser, therefore I have left it in `api`/`error` (instead of api/lexer/error).

**Comment by Nercury** [comment]
And then I _thought_ about it. If I am striving for concern separation I should keep lexer errors in lexer mod.

Now lexer error can be one of two things: `io::Error` or `twig::api::SyntaxError`.

**Comment by colin-kiegel** [comment]
Cool, I will have a look (later). :-)

File: src/api/lexer/lexing.rs [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+    /// Position of the next token to get.
+    next_pos: Position, // temporary field until I get cursor in.
+    env: &'i Lexer,
+    code: &'t str,
+}
+
+impl<'i, 't> Iterator for Tokens<'i, 't> {
+    type Item = LexingResult<ItemRef<'t>>;
+
+    fn next(&mut self) -> Option<LexingResult<ItemRef<'t>>> {
```
Comment (colin-kiegel) [inline src/api/lexer/lexing.rs:33]: TL;DR at the end ^^

Too bad rust does not have guaranteed tail call elimination. I did some investigations, what we could do instead.

First, I analyzed the nesting of sub-calls in my code and twigphp to look for any possibly infinite recursions.
The short answer is, there are definitely two possible recursion loops:
- state::Data::tokenize (aka lex_data) calls itself (luckily in a tail position)
- lex_expression calls itself (a) directly and possibly via (b) lex_expression -> lex_string -> lex_interpolation -> lex_expression

These recursions are separately closed in a sense they can be rewritten individually like

``` rust
'lex_data: loop {
    // do something
    if(/**/) { continue 'lex_data /* mimic tail-call */ }
    // ..
    if(/**/) { break 'lex_data /* or return */ }
    // ..
}
```

However there might be another recursion related to interpolation in double-quoted strings. This boils down to "can an expression/variable contain a string that contains a variable that contains another string that .."

Here are some details. I denote subcalls by "->", where tail-/non-tail-positions are marked with "*" and "_". No sub-calls: "()".

```
Initial -> *Data
Data -> *Data + *Final + _lex_comment + _lex_verbatim_data + _Block + _Expression
Final -> ()
lex_comment -> ()
lex_verbatim_data -> ()
Block -> *lex_expression
Expression/Variable -> *lex_expression
lex_expression -> *String + *PARENT (e.g. Block, Expression/Variable or Interpolation -> continue 'PARENT )
String -> *Interpolation
Interpolation -> *lex_expression (-> we can only reach this inside another lex_expression, so we could just `continue 'lex_expression` in our loop version - if it wasn't for the call from lex_expression to *PARENT -> this loop might be the most difficult one)
```

The good news is: sub-calls in non-tail-positions "_" don't create recursions (they only occur in Data, so this is easy to analyse).
- The lex_data-recursion is easy to rewrite (see above).
- And I am very confident the lex_expression recursion can be rewritten as a flat loop, too - however we probably want to share the code between Block and Expression/Variable.

We can't break a loop of the caller from inside a function - even a generic function is not flexible enough - but macros can. For readability I put the macros to the end - but for compilation they must be defined first.

``` rust
'lex_block: loop {
  // ..
  lex_expression!("block successfully parsed", break 'lex_block, continue 'lex_block); // not in tail position - but no guarantee needed, because not recursive
  continue 'lex_data // equivalent to a tail-call guarantee
  // ..
}

// shared code via macro - a bit nasty but it works(tm)
macro_rules! lex_expression {
    ( $message:expr, $break_parent:stmt, $continue_parent:stmt ) => {
        'lex_expression: loop {
        // do something
        println!($message);
        $break_parent;
        unreachable!(); // just to check the break_parent stmt really jumps somewhere else
        // ...
                // we could embedd the lex_string logic completely in this macro, or delegate to a sub-macro
        lex_string!("some argument", break 'lex_expression, continue 'lex_expression)
                //
        }
    };
}

// shared code via macro - a bit nasty but it works(tm)
macro_rules! lex_string {
    ( $message:expr, $break_parent:stmt, $continue_parent:stmt ) => {
        'lex_string: loop {
        // do something
        println!($message);
        $continue_parent;
        unreachable!(); // just to check the break_statement really jumps somewhere else
        // ...
        }
    };
}
```

I know this macro-thing probably feels a bit weird. But if we want to factor away most function calls in lexer, it seems possible. The only thing that needs a closer look is the possible expression-interpolation-recursion (this is where the stack could still blow up, if we can't reduce it to a flat loop). But at this point even I begin to question if it is really worth it... 

# TL;DR

So finally I arrive at the point to admit "let's just proceed with some enum-based match-branching in a loop - which would include the iterator pattern you are suggesting". ;-)

File: src/api/lexer/lexing.rs [hunk 11]
```diff
@@ -0,0 +1,83 @@
...
+pub struct ItemRef<'t> {
+    pub token: TokenRef<'t>,
+    pub position: Position,
+}
+
+/// Lexer token iterator.
+///
+/// 'i is iteration lifetime, or "one use of lexer".
+/// 't is template lifetime. It will live longer than this iteration.
+pub struct Tokens<'i, 't> {
```
Comment (colin-kiegel) [inline src/api/lexer/lexing.rs:23]: `Tokens`: It's only a detail - I prefer singular names like "Team" instead of "People", or "String" instead of "Characters",. For example `lexer::Job` would still work in this case - I think it would reflect very well that it is some intermediate thing that has a mutable internal state. We could also call it `lexer::Iter`. :-)

File: src/api/tokens/mod.rs [hunk 12]
```diff
@@ -0,0 +1,38 @@
+// This file is part of rust-web/twig
+//
+// For the copyright and license information, please view the LICENSE
+// file that was distributed with this source code.
+
+/*!
+Tokens, received from lexer output.
+*/
```
Comment (colin-kiegel) [inline src/api/tokens/mod.rs:8]: I read that multi-line comments are not rust-standard - but I can see the motivation to make life easier here. I plan to have a look at rustfmt. Maybe this can help with rust-like formatting and convert things like this? If not now - possibly in the future. :-)

Of course this is just reasoning about aesthetics. ^^

**Comment by colin-kiegel** [comment]
request to rename `mod tokens -> mod token`?
(I would prefer singular nouns as module names).

File: src/api/error/syntax.rs [hunk 13]
```diff
@@ -0,0 +1,44 @@
...
+
+use api::Position;
+use api::tokens::{ Token };
+
+#[derive(Debug)]
+pub struct SyntaxError {
+    pub code: SyntaxErrorCode,
+    pub starts_at: Position,
+    pub ends_at: Option<Position>,
+}
```
Comment (colin-kiegel) [inline src/api/error/syntax.rs:14]: ok :-)

File: src/api/lexer/lexing.rs (hunk 10 above)
Comment (Nercury) [inline src/api/lexer/lexing.rs:33]: OK, but we can definitely do it in a way that does not block future improvements.

File: src/api/tokens/mod.rs (hunk 12 above)
Comment (Nercury) [inline src/api/tokens/mod.rs:8]: They tried to remove them without success several times - I think at this point it is unlikely to happen :D

**Comment by Nercury** [comment]
> request to rename mod tokens -> mod token?

Yeah, that's just a detail.

File: src/api/error/syntax.rs (hunk 13 above)
Comment (Nercury) [inline src/api/error/syntax.rs:14]: I decided to raise the bar up to Rust level when it comes to error location display :)

File: src/api/lexer/error.rs [hunk 14]
```diff
@@ -0,0 +1,28 @@
...
+    }
+}
+
+impl From<SyntaxError> for LexingError {
+    fn from(other: SyntaxError) -> LexingError {
+        LexingError::Syntax(other)
+    }
+}
+
+pub type LexingResult<T> = result::Result<T, LexingError>;
```
Comment (colin-kiegel) [inline src/api/lexer/error.rs:28]: with current error management this should become `result::Result<T, Traced<LexingError>>;`

File: src/api/lexer/lexing.rs [hunk 15]
```diff
@@ -0,0 +1,83 @@
...
+impl<'i, 't> Tokens<'i, 't> {
+    pub fn new<'ii, 'tt>(lexer: &'ii Lexer, code: &'tt str) -> Tokens<'ii, 'tt> {
+        Tokens {
+            next_pos: Position { line: 1, column: 1 },
+            env: lexer,
+            code: code,
+        }
+    }
+
+    pub fn expect(&mut self, expected: TokenRef<'t>) -> LexingResult<TokenRef<'t>> {
```
Comment (colin-kiegel) [inline src/api/lexer/lexing.rs:59]: in my code I tried to be a bit more abstract ~ like `expect<T: pattern>' where both Token and TokenType implement`pattern`. This might help to match _any number_, without specifying the value.

**Comment by colin-kiegel** [comment]
I moved the generic error stuff to `src/api/error/*` (extension-dev perspective). I would like to try to separate abstract error wrappers/helpers a bit. Can we relocate the syntax error somewhere else? I think it belongs to lexer or lexing `src/api/lexer/error`.

Currently I imagine something like `src/error.rs` to define a `Traced<TwigError>`, where

```
pub enum TwigError {
    LexingError(LexingError),
    // ...
}
```

Where the Display + Debug traits should only add a thin layer of information - and we can still discuss and optimize the order of appearance on the screen. A user calling functions on the `facade` should always receive a `Traced<TwigError>`. If they need to destructure this, they should start with `TwigError`.

We can also think about re-exporting all specific errors in `src/error.rs` - this would be a clear separation from abstract helpers in `src/api/error.rs`

**Comment by colin-kiegel** [comment]
PS: Here is what I mean (from #12 `src/engine/error.rs`).

```
#[derive(Debug)]
pub enum TwigError {
    Lexer(LexerError),
    // ...
}

impl From<LexerError> for TwigError {
    fn from(err: LexerError) -> TwigError {
        TwigError::Lexer(err)
 }}

impl Error for TwigError {
    fn description(&self) -> &str {
        match *self {
            TwigError::Lexer(..) => "Twig lexer error.",
            // ...
}}}

impl Display for TwigError {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        try!(write!(f, "{}", self.description()));

        match *self {
            TwigError::Lexer(ref e) => e.fmt(f),
            // ...
}}}
```

**Comment by Nercury** [comment]
Ok, looks great, I will do that.

**Comment by Nercury** [comment]
Continued in #21.
